api/authenticator.py

```diff
@@ -369,53 +372,33 @@ class LibraryAuthenticator(object):
 
     @classmethod
     def from_config(cls, _db, library):
-        """Initialize an Authenticator from site configuration.
+        """Initialize an Authenticator for the given Library based on its
+        configured ExternalIntegrations.
         """
-        
-        authentication_policy = Configuration.policy("authentication")
-        if not authentication_policy:
-            raise CannotLoadConfiguration(
-                "No authentication policy given."
-            )
-        
-        if (not isinstance(authentication_policy, dict)
-            or not 'providers' in authentication_policy):
-            raise CannotLoadConfiguration(
-                "Authentication policy must be a dictionary with key 'providers'."
-            )
-        bearer_token_signing_secret = authentication_policy.get(
-            'bearer_token_signing_secret'
+
+        bearer_token_signing_secret = OAuthAuthenticationProvider.bearer_token_signing_secret(
+            _db
         )
-        providers = authentication_policy['providers']        
-        if isinstance(providers, dict):
-            # There's only one provider.
-            providers = [providers]
-            
-        # Start with an empty list of authenticators.        
+
+        # Start with an empty list of authenticators.
         authenticator = cls(
             _db=_db, library=library,
             bearer_token_signing_secret=bearer_token_signing_secret
         )
 
-        # Register each provider.
-        for provider_dict in providers:
-            if not isinstance(provider_dict, dict):
-                raise CannotLoadConfiguration(
-                    "Provider %r is invalid; must be a dictionary." %
-                    provider_dict
-                )
-            authenticator.register_provider(provider_dict)
+        # Find all of this library's ExternalIntegrations set up with
+        # the goal of authenticating patrons.
+        integrations = _db.query(ExternalIntegration).join(
+            ExternalIntegration.libraries).filter(
+            ExternalIntegration.goal==ExternalIntegration.PATRON_AUTH_GOAL
+        ).filter(
+            Library.id==library.id
+        )
+        # Turn each such ExternalIntegration into an
+        # AuthenticationProvider.
+        for integration in integrations:
+            authenticator.register_provider(integration)
                 
-        if (not authenticator.basic_auth_provider
-            and not authenticator.oauth_providers_by_name):
-            # TODO: This isn't unacceptable: a fully open-access
-            # collection doesn't need any authentication providers.
-            # But supporting that case requires specialized work, e.g.
-            # getting rid of all the links to controllers that require
-            # authentication.
-            raise CannotLoadConfiguration(
-                "No authentication provider configured"
-            )
```

Comment:
I had to take this code out, even though we still have those links to controllers, because when you've just created a library, there's going to be a period where your library has no authentication providers, and we need to treat that as a valid configuration, if only so the admin can use the interface to add a provider.